Provide API to send ICMP Echo-Request messages

jukkar opened the pull request:
This creates a public ICMP API so that applications can send Echo-Request messages. This was allowed earlier if application used network internal header files but this was more of a kludge than a proper feature.

This change also allows offloaded pings to be sent. Currently the unit tests show how they could be used.

The network internals are changed to use also this new API so the private and legacy ICMP API is removed.

jukkar committed:
net: icmp: Introduce new API to send ICMP messages

Allow user to send ICMP Echo Request message a.k.a pings.
The same ICMP API and framework is used for network stack
internal needs in later commits. One benefit for this new
API is that it allows sending Echo Requests to offlined
network devices.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Use the ICMP API for receiving ICMP messages

Modify the internal network stack code to use the ICMP API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Use the ICMP API for ping command

Change the ping command to use the ICMP API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
@jtbaumann / @rtalbott-tmo I am trying to figure out the offloaded ping API. Sending offloaded ping should be easy, we just call in net stack some function the driver has registered and then that func can do the actual sending.
Is it possible to receive a ping response back i.e., can driver call some function in net stack when the driver receives a ping response. Without the response, the "net ping" command is not working properly as it thinks the responses are lost.

jukkar committed:
net: icmp: Add offloading support

Add suitable functions that offloaded network device driver
can receive ICMP Echo-Request (ping) requests and give
ICMP responses back when it receives them from the network.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on include/zephyr/net/icmp.h:
```diff
@@ -0,0 +1,149 @@
+ *        to this Echo Request. May be zero.
+ * @param tc_tos Can be either IPv4 Type-of-service field value,
+ *        or IPv6 Traffic Class field value. Represents combined DSCP and
+ *        ECN values.
```
`priority` doc is missing

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,390 @@
+#if defined(CONFIG_NET_IPV6)
+#if defined(MY_LOG_LEVEL)
+#if MY_LOG_LEVEL < CONFIG_NET_ICMPV6_LOG_LEVEL
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
```
Shouldn't `MY_LOG_LEVEL` be undefined before? Looks like symbol redefinition right now.

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,390 @@
+	}
+
+	/* Take the first address of the network interface */
+	src = &iface->config.ip.ipv4->unicast[0].address.in_addr;
```
I see it's how it is implemented right now, but wouldn't it be better to use `net_if_ipv4_select_src_addr()` here?

jukkar replied:
Yes, will fix this and the other issues reported by twister in next version. I have also proposal for the offloaded API in next version. I will also prepare test cases for this.

rlubos reviewed (commented):
Looks pretty reasonable. Indeed it should facilitate ICMP handling for offloaded interafces, just need to figure out ICMP API for the offloaded case.

jukkar committed:
tests: net: icmp: Add test cases for new ICMP API

Tests cases for new ICMP API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: icmp: Add offloaded ICMP API tests

Add some tests that verify that offloaded ICMP API works
as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Added proposal for offloaded ping API
* Added simple non-offloaded test which work ok.
* Added offloaded test which are still WIP

jukkar committed:
tests: net: icmpv4: Convert to use new ICMP API

Converting the ICMPv4 tests to use the new ICMP API so that
the tests pass.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on include/zephyr/net/icmp.h:
```diff
@@ -46,6 +46,44 @@ typedef int (*net_icmp_handler_t)(struct net_icmp_ctx *ctx,
+ */
+typedef int (*net_icmp_offload_ping_handler_t)(struct net_icmp_ctx *ctx,
+					       struct net_if *iface,
+					       struct sockaddr *dst,
+					       uint16_t identifier,
+					       uint16_t sequence,
+					       uint8_t tc_tos,
+					       int priority,
+					       const void *data,
+					       size_t data_size,
```
Perhaps it'd make sense to define `struct net_icmp_ping_params` and use it as a function argument instead of repeating the lengthy parameter list?

jukkar replied:
Yes, makes sense. I will fix this.

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -42,6 +42,10 @@ LOG_MODULE_REGISTER(net_icmp, MY_LOG_LEVEL);
 static K_MUTEX_DEFINE(lock);
 static sys_slist_t handlers;
 
+#if defined(CONFIG_NET_OFFLOAD)
```
We should likely support socket offloading as well (` || defined(CONFIG_NET_SOCKETS_OFFLOAD)`), here and in other places.

jukkar replied:
Sure, I had that in pipeline, the code was just not yet having that support.

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -358,6 +384,45 @@ static int send_icmpv6_echo_request(struct net_icmp_ctx *ctx,
+		return -EINVAL;
+	}
+
+	if (!net_if_is_ip_offloaded(iface)) {
```
We have a helper function that covers both offloading scenarios (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_if.c#L4247), perhaps we should make it public and use it instead?

jukkar replied:
Yes, makes sense, will do.

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -358,6 +384,45 @@ static int send_icmpv6_echo_request(struct net_icmp_ctx *ctx,
+#if defined(CONFIG_NET_OFFLOAD)
+	SYS_SLIST_FOR_EACH_CONTAINER(&offload_handlers, offload, node) {
+		if (offload->iface == iface) {
+			ping_handler = &offload->ping_handler;
```
Hmm, we overwrite local variable (pointer) here, shouldn't it actually be dereferenced before write?

jukkar replied:
Hmm, I am not following you here. We are returning the ping_handler to the caller so that is why the code looks like this.

rlubos replied:
I think we're on the same page now, the latest push illustrates best what I meant here:
```patch
-			ping_handler = &offload->ping_handler;
+			*ping_handler = offload->ping_handler;
```

jukkar replied:
Ah yes, I had fixed that earlier before your comment but before sending that version to review so got a bit confused.

jukkar committed:
tests: net: icmpv6: Convert to use new ICMP API

Converting the ICMPv6 tests to use the new ICMP API so that
the tests pass.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Add function for network interface offloading status

Add new net_if_is_offloaded(iface) that returns true if
the network interface is offloaded (either IP or socket
offloading is enabled for that interface) and false if
the interface is not offloaded.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: zperf: Fix the IPv6 ping done in shell

The zperf shell sends a IPv6 ping at the start when working
with IPv6. Convert the sending of the ping to use the new API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv6_fragment: Fix ICMP usage in the test

Convert the test to use new ICMP API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: mld: Fix ICMP usage in the test

Convert the test to use new ICMP API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,479 @@
+static int send_icmpv4_echo_request(struct net_icmp_ctx *ctx,
+				    struct net_if *iface,
+				    struct in_addr *dst,
+				    uint16_t identifier,
+				    uint16_t sequence,
+				    uint8_t tos,
+				    int priority,
+				    const void *data,
+				    size_t data_size,
```
Maybe this internal function could accept `struct net_icmp_ping_params*` as well?

jukkar replied:
Sure, np.

rlubos commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,479 @@
+static int send_icmpv6_echo_request(struct net_icmp_ctx *ctx,
+				    struct net_if *iface,
+				    struct in6_addr *dst,
+				    uint16_t identifier,
+				    uint16_t sequence,
+				    uint8_t tc,
+				    int priority,
+				    const void *data,
+				    size_t data_size,
```
Ditto

rlubos reviewed (dismissed):
Looks good, just one minor suggestion.

rlubos reviewed (dismissed):
LGTM

pdgendt commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,596 @@
+#include "net_stats.h"
+
+static K_MUTEX_DEFINE(lock);
+static sys_slist_t handlers;
+
+#if defined(CONFIG_NET_OFFLOADING_SUPPORT)
+static sys_slist_t offload_handlers;
```
There is a macro for static initialization of a single-linked list.

```suggestion
static sys_slist_t handlers = SYS_SLIST_STATIC_INIT(&handlers);

#if defined(CONFIG_NET_OFFLOADING_SUPPORT)
static sys_slist_t offload_handlers = SYS_SLIST_STATIC_INIT(&offload_handlers);
```

jukkar replied:
Good point, will fix.

pdgendt commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,596 @@
+
+#if !defined(MY_LOG_LEVEL)
+#define MY_LOG_LEVEL LOG_LEVEL_INF
+#endif
```
This looks a bit messy, I wonder if you can use `COND_CODE_1`:

```c
#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV6)
#define ICMP_LOG_LEVEL COND_CODE_1(CONFIG_NET_ICMPV4_LOG_LEVEL > CONFIG_NET_ICMPV6_LOG_LEVEL, \
	(CONFIG_NET_ICMPV4_LOG_LEVEL), (CONFIG_NET_ICMPV6_LOG_LEVEL))

#elif defined(CONFIG_NET_IPV4)
#define ICMP_LOG_LEVEL CONFIG_NET_ICMPV4_LOG_LEVEL
#elif defined(CONFIG_NET_IPV6)
#define ICMP_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
#else
#define ICMP_LOG_LEVEL LOG_LEVEL_INF
#endif
```

jukkar replied:
Cool, I will try that.

pdgendt replied:
It looks like it doesn't though :(

pdgendt commented on subsys/net/ip/icmp.c:
```diff
@@ -0,0 +1,589 @@
+
+/* Use highest log level if both IPv4 and IPv6 are defined */
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV6)
+#define ICMP_LOG_LEVEL COND_CODE_1(CONFIG_NET_ICMPV4_LOG_LEVEL > CONFIG_NET_ICMPV6_LOG_LEVEL, \
+	(CONFIG_NET_ICMPV4_LOG_LEVEL), (CONFIG_NET_ICMPV6_LOG_LEVEL))
```
This doesn't appear to work after some testing, an additional `#if` will be required:

```c
#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV6)

#if CONFIG_NET_ICMPV4_LOG_LEVEL > CONFIG_NET_ICMPV6_LOG_LEVEL
#define ICMP_LOG_LEVEL CONFIG_NET_ICMPV4_LOG_LEVEL
#else
#define ICMP_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
#endif

#elif defined(CONFIG_NET_IPV4)
#define ICMP_LOG_LEVEL CONFIG_NET_ICMPV4_LOG_LEVEL
#elif defined(CONFIG_NET_IPV6)
#define ICMP_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
#else
#define ICMP_LOG_LEVEL LOG_LEVEL_INF
#endif
```

jukkar replied:
It looked very cool, well back to normal checks then :(
